File: 2-add-two-numbers/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  const addTwoNumbersHelper = (tup1, tup2, regroup) => {
    if (!tup1 || !tup2) {
      if (!tup1 && !tup2) {
        return regroup ? new ListNode(1, null) : null
      } else if (!tup1) {
        if (regroup) {
          const result = tup2.val + regroup
          return new ListNode(result%10, addTwoNumbersHelper(null, tup2.next, ~~(result/10)))
        } else {
          return tup2
        }
      } else {
        return addTwoNumbersHelper(tup2, tup1, regroup)
      }
    } else {
      const result = tup1.val + tup2.val + regroup
      return new ListNode(result%10, addTwoNumbersHelper(tup1.next, tup2.next, ~~(result/10)))
    }
  }
  
  return addTwoNumbersHelper(l1, l2, 0)
}

File: 2-add-two-numbers/2-add-two-numbers.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    val dummy = ListNode(-1)
    var p1 = l1
    var p2 = l2
    var p = dummy
    var c = 0
    while (p1 != null || p2 != null || c > 0) {
      var sum = c
      if (p1 != null) {
        sum += p1.x
        p1 = p1.next
      }
      if (p2 != null) {
        sum += p2.x
        p2 = p2.next
      }
      p.next = ListNode(sum % 10)
      c = sum / 10
      p = p.next
    }
    return dummy.next
  }
}